core/sail/lmdb/src/main/java/org/eclipse/rdf4j/sail/lmdb/LmdbRecordIterator.java

```diff
@@ -126,73 +126,83 @@ class LmdbRecordIterator implements RecordIterator {
 
 	@Override
 	public long[] next() {
-		long stamp = txnLock.readLock();
-		try {
-			if (closed) {
-				log.debug("Calling next() on an LmdbRecordIterator that is already closed, returning null");
-				return null;
+		long stamp = txnLock.tryOptimisticRead();
+		long[] value = nextInternal();
```

Comment:
@kenwenzel Is the read/write locking needed as a protection for LMDB, or is it safe to concurrently do all the operations in `nextInternal()` while there is a write?

Reply:
@hmottestad It's only required to protect txnRef itself to not accidentally close read transactions during a read operation. As far as I remember this is required to correctly support autogrow which needs to close all transactions.

Reply:
So if autogrow happens in the middle of a read operation that doesn't hold the lock, what will happens with the read operation?

Reply:
It will fail as the transaction will be closed.